config/src/test/java/org/springframework/security/config/annotation/web/configuration/sec2377/Sec2377Tests.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.config.annotation.web.configuration.sec2377;
+
+import org.junit.Test;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.web.context.support.AnnotationConfigWebApplicationContext;
+
+/**
+ * @author Rob Winch
+ * @author Josh Cummings
+ */
+public class Sec2377Tests {
+
+	@Test
+	public void contextRefreshWhenParentAndChildRegisteredThenNoException() {
+		AnnotationConfigWebApplicationContext parent =
```

Comment:
We should use `SpringTestRule` for all tests. For example: 

```
	@Rule
	public final SpringTestRule spring = new SpringTestRule();

this.spring.register(Sec2377AConfig.class).autowire();
```

Search for examples of `SpringTestRule`

Reply:
To represent the scenario correctly, I need two contexts, correct? If I've got that right, and if I'm reading `SpringTestRule` correctly, then I'll need two instances of the rule as well as add a `parent` method to tie one context to the other through `SpringTestRule`, e.g.:

```
@Rule
public final SpringTestRule parent = new SpringTestRule();

@Rule
public final SpringTestRule child = new SpringTestRule();

@Test
public void test() {
  // something like this
  parent.register(Sec2377AConfig.class).autowire();
  child.register(Sec2377BConfig.class).parent(parent).autowire();
}
```

Is this the kind of thing you are expecting?

Reply:
That looks right with one correction

`this.child.register(Sec2377BConfig.class).getContext().setParent(parent.getContext()).autowire();`